File: Assignments/ProgrammingAssignment01/OS/exec.c
#include "types.h"
#include "param.h"
#include "defs.h"
#include "memlayout.h"
#include "mmu.h"
#include "proc.h"
#include "elf.h"
#include "arm.h"

// -------------------------------------------------------------
// exec(path, argv)
//  - Replaces the current process image with a new program.
//  - Loads 'path' (like "/ls") into memory, sets up stack with argv[],
//    and jumps to the program's entry point defined in its ELF header.
// -------------------------------------------------------------
int exec(char *path, char **argv)
{
    struct elfhdr elf;          // ELF header (holds entry point, prog headers)
    struct inode *ip;           // inode for the program file
    struct proghdr ph;          // program header (describes segments to load)
    pde_t *pgdir;               // new page directory for process
    pde_t *oldpgdir;            // old page directory (to free later)
    char *s, *last;             // helpers for program name
    int i, off;                 
    uint argc;                  // argument count
    uint sz;                    // size of process memory (bytes)
    uint sp;                    // stack pointer
    uint ustack[3 + MAXARG + 1]; // temporary array to build user stack

    // ---------------------------------------------------------
    // Step 1: Locate the program file in the file system.
    // ---------------------------------------------------------
    if ((ip = namei(path)) == 0) {   // find inode for "path"
        return -1;                   // file not found
    }
    ilock(ip);                       // lock the inode for safe read

    // ---------------------------------------------------------
    // Step 2: Validate ELF header.
    // ---------------------------------------------------------
    if (readi(ip, (char*) &elf, 0, sizeof(elf)) < sizeof(elf)) {
        goto bad;    // couldn’t read ELF header
    }
    if (elf.magic != ELF_MAGIC) {    // check ELF magic number
        goto bad;    // not a valid ELF binary
    }

    pgdir = 0;

    // ---------------------------------------------------------
    // Step 3: Allocate a new page directory (new virtual memory space).
    // ---------------------------------------------------------
    if ((pgdir = kpt_alloc()) == 0) {
        goto bad;    // failed to allocate kernel page table
    }

    // ---------------------------------------------------------
    // Step 4: Load program segments into memory.
    // ---------------------------------------------------------
    sz = 0;                         // start process size at 0
    for (i = 0, off = elf.phoff;    // loop through program headers
         i < elf.phnum;
         i++, off += sizeof(ph)) 
    {
        // read program header i
        if (readi(ip, (char*) &ph, off, sizeof(ph)) != sizeof(ph)) {
            goto bad;
        }

        if (ph.type != ELF_PROG_LOAD) {
            continue;  // only load segments marked "LOAD"
        }

        if (ph.memsz < ph.filesz) {  // sanity check
            goto bad;
        }

        // extend process memory to hold this segment
        if ((sz = allocuvm(pgdir, sz, ph.vaddr + ph.memsz)) == 0) {
            goto bad;
        }

        // copy the segment from file into virtual memory
        if (loaduvm(pgdir, (char*) ph.vaddr, ip, ph.off, ph.filesz) < 0) {
            goto bad;
        }
    }

    iunlockput(ip);   // release inode
    ip = 0;           // prevent accidental reuse

    // ---------------------------------------------------------
    // Step 5: Allocate stack (two pages).
    //   - First page: guard (inaccessible)
    //   - Second page: actual user stack
    // ---------------------------------------------------------
    sz = align_up(sz, PTE_SZ);   // round size up to page boundary
    if ((sz = allocuvm(pgdir, sz, sz + 2 * PTE_SZ)) == 0) {
        goto bad;
    }
    clearpteu(pgdir, (char*) (sz - 2 * PTE_SZ)); // mark guard page as invalid

    sp = sz;    // user stack pointer starts at end of stack region

    // ---------------------------------------------------------
    // Step 6: Push arguments (argv[]) onto the new user stack.
    // ---------------------------------------------------------
    for (argc = 0; argv[argc]; argc++) {
        if (argc >= MAXARG) {
            goto bad;  // too many arguments
        }

        // move stack pointer down to allocate space for string
        sp = (sp - (strlen(argv[argc]) + 1)) & ~3; // align to 4 bytes

        // copy argument string from kernel -> user memory
        if (copyout(pgdir, sp, argv[argc], strlen(argv[argc]) + 1) < 0) {
            goto bad;
        }

        ustack[argc] = sp;   // record pointer to this argument
    }
    ustack[argc] = 0;        // argv[argc] = NULL (terminator)

    // In ARM ABI: 
    // r0 = argc, r1 = argv
    proc->tf->r0 = argc;
    proc->tf->r1 = sp - (argc + 1) * 4; // where argv[] will be copied

    // allocate space on user stack for argv[]
    sp -= (argc + 1) * 4;
    if (copyout(pgdir, sp, ustack, (argc + 1) * 4) < 0) {
        goto bad;
    }

    // ---------------------------------------------------------
    // Step 7: Save process name (for debugging/ps).
    // ---------------------------------------------------------
    for (last = s = path; *s; s++) {
        if (*s == '/') {      // skip to last part of path
            last = s + 1;
        }
    }
    safestrcpy(proc->name, last, sizeof(proc->name));

    // ---------------------------------------------------------
    // Step 8: Commit to the new image.
    // ---------------------------------------------------------
    oldpgdir = proc->pgdir;    // remember old memory
    proc->pgdir = pgdir;       // install new page directory
    proc->sz = sz;             // set new memory size
    proc->tf->pc = elf.entry;  // set program counter to ELF entry point
    proc->tf->sp_usr = sp;     // set user stack pointer
    switchuvm(proc);           // switch to new address space
    freevm(oldpgdir);          // free old memory

    return 0;  // success!

// -------------------------------------------------------------
// Failure path: free resources and return -1
// -------------------------------------------------------------
bad:
    if (pgdir) {
        freevm(pgdir);
    }
    if (ip) {
        iunlockput(ip);
    }
    return -1;
}

File: xv6-rust/src/exec.rs
use crate::types_h::*;
use crate::param_h::*;
// use crate::defs::*;
use crate::memlayout_h::*;
use crate::mmu_h::*;
use crate::proc_h::*;
use crate::elf_h::*;
use crate::arm_h::*;

pub unsafe fn exec(path: *const u8, argv: *const *const u8) -> i32 {
    let mut elf: ElfHdr = core::mem::zeroed();
    let mut ip: *mut Inode = core::ptr::null_mut();
    let mut ph: ProgHdr = core::mem::zeroed();
    let mut pgdir: *mut Pde = core::ptr::null_mut();
    let mut oldpgdir: *mut Pde = core::ptr::null_mut();
    let mut s: *const u8;
    let mut last: *const u8;
    let mut i: usize;
    let mut off: usize;
    let mut argc: usize = 0;
    let mut sz: usize = 0;
    let mut sp: usize;
    let mut ustack: [usize; 3 + MAXARG + 1] = [0; 3 + MAXARG + 1];

    // Step 1: Locate program file
    ip = namei(path);
    if ip.is_null() {
        return -1;
    }
    ilock(ip);

    // Step 2: Validate ELF header
    if readi(ip, &mut elf as *mut _ as *mut u8, 0, core::mem::size_of::<ElfHdr>()) 
        < core::mem::size_of::<ElfHdr>() as i32 
    {
        goto_bad!(pgdir, ip);
    }

    if elf.magic != ELF_MAGIC {
        goto_bad!(pgdir, ip);
    }

    pgdir = kpt_alloc();
    if pgdir.is_null() {
        goto_bad!(pgdir, ip);
    }

    // Step 4: Load program segments
    sz = 0;
    i = 0;
    off = elf.phoff as usize;

    while i < elf.phnum as usize {
        if readi(ip, &mut ph as *mut _ as *mut u8, off, core::mem::size_of::<ProgHdr>()) 
            != core::mem::size_of::<ProgHdr>() as i32 
        {
            goto_bad!(pgdir, ip);
        }

        if ph.type_ != ELF_PROG_LOAD {
            i += 1;
            off += core::mem::size_of::<ProgHdr>();
            continue;
        }

        if ph.memsz < ph.filesz {
            goto_bad!(pgdir, ip);
        }

        if (sz = allocuvm(pgdir, sz, (ph.vaddr + ph.memsz) as usize)) == 0 {
            goto_bad!(pgdir, ip);
        }

        if loaduvm(pgdir, ph.vaddr as *mut u8, ip, ph.off as usize, ph.filesz as usize) < 0 {
            goto_bad!(pgdir, ip);
        }

        i += 1;
        off += core::mem::size_of::<ProgHdr>();
    }

    iunlockput(ip);
    ip = core::ptr::null_mut();

    // Step 5: Allocate stack (2 pages)
    sz = align_up(sz, PTE_SZ);
    if (sz = allocuvm(pgdir, sz, sz + 2 * PTE_SZ)) == 0 {
        goto_bad!(pgdir, ip);
    }
    clearpteu(pgdir, (sz - 2 * PTE_SZ) as *mut u8); // guard page

    sp = sz;

    // Step 6: Push arguments
    while !(*argv.add(argc)).is_null() {
        if argc >= MAXARG {
            goto_bad!(pgdir, ip);
        }

        sp = (sp - strlen(*argv.add(argc)) - 1) & !3; // align to 4 bytes

        if copyout(pgdir, sp, *argv.add(argc), strlen(*argv.add(argc)) + 1) < 0 {
            goto_bad!(pgdir, ip);
        }

        ustack[argc] = sp;
        argc += 1;
    }

    ustack[argc] = 0;

    proc.tf.as_mut().unwrap().r0 = argc;
    proc.tf.as_mut().unwrap().r1 = sp - (argc + 1) * 4;

    sp -= (argc + 1) * 4;
    if copyout(pgdir, sp, ustack.as_ptr() as *const u8, (argc + 1) * 4) < 0 {
        goto_bad!(pgdir, ip);
    }

    // Step 7: Save process name
    last = path;
    s = path;
    while *s != 0 {
        if *s == b'/' {
            last = s.add(1);
        }
        s = s.add(1);
    }
    safestrcpy(proc.name.as_mut_ptr(), last, core::mem::size_of::<[u8; 16]>());

    // Step 8: Commit new image
    oldpgdir = proc.pgdir;
    proc.pgdir = pgdir;
    proc.sz = sz;
    proc.tf.as_mut().unwrap().pc = elf.entry as usize;
    proc.tf.as_mut().unwrap().sp_usr = sp;

    switchuvm(proc);
    freevm(oldpgdir);

    return 0;

    macro_rules! goto_bad {
        ($pgdir:expr, $ip:expr) => {{
            if !$pgdir.is_null() {
                freevm($pgdir);
            }
            if !$ip.is_null() {
                iunlockput($ip);
            }
            return -1;
        }};
    }
}
